In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
visa_df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified  
2      122996.8600         Year                  Y      Denied  
3       83434.0300         Year                  Y      Denied  
4      149907.3900         Year                  Y   Certified

In [5]:
visa_df.select_dtypes(include='object').columns

Index(['case_id', 'continent', 'education_of_employee', 'has_job_experience',
       'requires_job_training', 'region_of_employment', 'unit_of_wage',
       'full_time_position', 'case_status'],
      dtype='object')

**Categorical To Numeric**

- map method
- where method
- one hot encoder
- Label Encoder

**Map Method**

In [9]:
unique=visa_df['case_status'].unique()
d={}
p=0
for i in unique:
    d[i]=p
    p+=1
d

{'Denied': 0, 'Certified': 1}

$So Now with Map method$

In [13]:
visa_df['case_status_new']=visa_df['case_status'].map(d)
visa_df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  \
0         592.2029         Hour                  Y      Denied   
1       83425.6500         Year                  Y   Certified   
2      122996.8600         Year                  Y      Denied   
3       83434.0300         Year                  Y      Denied   
4      149907.3900         Year                  Y   Certified   

   case_status_new  
0                0  
1                1  
2                0  
3                0  
4                1

In [15]:
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
visa_df['case_status']=visa_df['case_status'].map(d)
visa_df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position  case_status  
0         592.2029         Hour                  Y            0  
1       83425.6500         Year                  Y            1  
2      122996.8600         Year                  Y            0  
3       83434.0300         Year                  Y            0  
4      149907.3900         Year                  Y            1

In [18]:
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
d1={}
for i in visa_df.select_dtypes(include='object').columns:
    unique=visa_df[i].unique()
    p=0
    for j in unique:
        d1[j]=p
        p+=1
    visa_df[i]=visa_df[i].map(d1)
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          0                      0                   0   
1            1          0                      1                   1   
2            2          0                      2                   0   
3            3          0                      2                   0   
4            4          1                      1                   1   
...        ...        ...                    ...                 ...   
25475    25475          0                      2                   1   
25476    25476          0                      0                   1   
25477    25477          0                      1                   1   
25478    25478          0                      1                   1   
25479    25479          0                      2                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         0         592.2029             0   
1                         1       83425.6500             1   
2                         0      122996.8600             1   
3                         0       83434.0300             1   
4                         2      149907.3900             1   
...                     ...              ...           ...   
25475                     2       77092.5700             1   
25476                     1      279174.7900             1   
25477                     2      146298.8500             1   
25478                     0       86154.7700             1   
25479                     3       70876.9100             1   

       full_time_position  case_status  
0                       0            0  
1                       0            1  
2                       0            0  
3                       0            0  
4                       0            1  
...                   ...          ...  
25475                   0            1  
25476                   0            1  
25477                   1            1  
25478                   0            1  
25479                   0            1  

[25480 rows x 12 columns]

**Label Encoder**

In [20]:
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
visa_df['case_status']=le.fit_transform(visa_df['case_status'])
visa_df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position  case_status  
0         592.2029         Hour                  Y            1  
1       83425.6500         Year                  Y            0  
2      122996.8600         Year                  Y            1  
3       83434.0300         Year                  Y            1  
4      149907.3900         Year                  Y            0

In [21]:
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
for i in visa_df.select_dtypes(include='object').columns:
    visa_df[i]=le.fit_transform(visa_df[i])
visa_df.head()

case_id  continent  education_of_employee  has_job_experience  \
0        0          1                      2                   0   
1        1          1                      3                   1   
2        2          1                      0                   0   
3        3          1                      0                   0   
4        4          0                      3                   1   

   requires_job_training  no_of_employees  yr_of_estab  region_of_employment  \
0                      0            14513         2007                     4   
1                      0             2412         2002                     2   
2                      1            44444         2008                     4   
3                      0               98         1897                     4   
4                      0             1082         2005                     3   

   prevailing_wage  unit_of_wage  full_time_position  case_status  
0         592.2029             0                   1            1  
1       83425.6500             3                   1            0  
2      122996.8600             3                   1            1  
3       83434.0300             3                   1            1  
4      149907.3900             3                   1            0

**Label Encoder is famous in Industry and also Easy**

**Where Method**

3 arguments

- condition
- True Value
- False Value

In [22]:
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
visa_df['case_status']=np.where(visa_df['case_status']=='Certified',0,1)
visa_df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position  case_status  
0         592.2029         Hour                  Y            1  
1       83425.6500         Year                  Y            0  
2      122996.8600         Year                  Y            1  
3       83434.0300         Year                  Y            1  
4      149907.3900         Year                  Y            0

$$USed only for Binary, means 2 values only$$

**One Hot Encoder**

- on represents with 1, off represents with 0
- Example: Case_Status Column
  - Denied
  - Certified
- Has 2 unique columns
- So, If we apply One Hot Encoder on case_status column it will creare 2 extra columns
  - case_status_Denied
  - case_status_Certified


In [23]:
import pandas as pd
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
pd.get_dummies(visa_df['case_status'],dtype='int')

Certified  Denied
0              0       1
1              1       0
2              0       1
3              0       1
4              1       0
...          ...     ...
25475          1       0
25476          1       0
25477          1       0
25478          1       0
25479          1       0

[25480 rows x 2 columns]

In [24]:
import pandas as pd
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
pd.get_dummies(visa_df,dtype='int')

no_of_employees  yr_of_estab  prevailing_wage  case_id_EZYV01  \
0                14513         2007         592.2029               1   
1                 2412         2002       83425.6500               0   
2                44444         2008      122996.8600               0   
3                   98         1897       83434.0300               0   
4                 1082         2005      149907.3900               0   
...                ...          ...              ...             ...   
25475             2601         2008       77092.5700               0   
25476             3274         2006      279174.7900               0   
25477             1121         1910      146298.8500               0   
25478             1918         1887       86154.7700               0   
25479             3195         1960       70876.9100               0   

       case_id_EZYV02  case_id_EZYV03  case_id_EZYV04  case_id_EZYV05  \
0                   0               0               0               0   
1                   1               0               0               0   
2                   0               1               0               0   
3                   0               0               1               0   
4                   0               0               0               1   
...               ...             ...             ...             ...   
25475               0               0               0               0   
25476               0               0               0               0   
25477               0               0               0               0   
25478               0               0               0               0   
25479               0               0               0               0   

       case_id_EZYV06  case_id_EZYV07  ...  region_of_employment_South  \
0                   0               0  ...                           0   
1                   0               0  ...                           0   
2                   0               0  ...                           0   
3                   0               0  ...                           0   
4                   0               0  ...                           1   
...               ...             ...  ...                         ...   
25475               0               0  ...                           1   
25476               0               0  ...                           0   
25477               0               0  ...                           1   
25478               0               0  ...                           0   
25479               0               0  ...                           0   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                              1                  1                   0   
1                              0                  0                   0   
2                              1                  0                   0   
3                              1                  0                   0   
4                              0                  0                   0   
...                          ...                ...                 ...   
25475                          0                  0                   0   
25476                          0                  0                   0   
25477                          0                  0                   0   
25478                          1                  0                   0   
25479                          0                  0                   0   

       unit_of_wage_Week  unit_of_wage_Year  full_time_position_N  \
0                      0                  0                     0   
1                      0                  1                     0   
2                      0                  1                     0   
3                      0                  1                     0   
4                      0                  1                     0   
...                  ...                ...                  

- So 25480 rows into 25510 columns, which is too much
- Note:
  - Drop the id column

In [27]:
import pandas as pd
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
visa_df.drop('case_id',axis=1,inplace=True)
pd.get_dummies(visa_df)

no_of_employees  yr_of_estab  prevailing_wage  continent_Africa  \
0                14513         2007         592.2029             False   
1                 2412         2002       83425.6500             False   
2                44444         2008      122996.8600             False   
3                   98         1897       83434.0300             False   
4                 1082         2005      149907.3900              True   
...                ...          ...              ...               ...   
25475             2601         2008       77092.5700             False   
25476             3274         2006      279174.7900             False   
25477             1121         1910      146298.8500             False   
25478             1918         1887       86154.7700             False   
25479             3195         1960       70876.9100             False   

       continent_Asia  continent_Europe  continent_North America  \
0                True             False                    False   
1                True             False                    False   
2                True             False                    False   
3                True             False                    False   
4               False             False                    False   
...               ...               ...                      ...   
25475            True             False                    False   
25476            True             False                    False   
25477            True             False                    False   
25478            True             False                    False   
25479            True             False                    False   

       continent_Oceania  continent_South America  \
0                  False                    False   
1                  False                    False   
2                  False                    False   
3                  False                    False   
4                  False                    False   
...                  ...                      ...   
25475              False                    False   
25476              False                    False   
25477              False                    False   
25478              False                    False   
25479              False                    False   

       education_of_employee_Bachelor's  ...  region_of_employment_South  \
0                                 False  ...                       False   
1                                 False  ...                       False   
2                                  True  ...                       False   
3                                  True  ...                       False   
4                                 False  ...                        True   
...                                 ...  ...                         ...   
25475                              True  ...                        True   
25476                             False  ...                       False   
25477                             False  ...                        True   
25478                             False  ...                       False   
25479                              True  ...                       False   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                           True               True               False   
1                          False              False               False   
2                           True              False               False   
3                           True              False               False   
4                          False              False               False   
...                          ...                ...                 ...   
25475                      False              False               False   
25476                      False              False               False   
25477                      False              False               False   
2

In [28]:
import pandas as pd
visa_df=pd.read_csv(r"C:\Users\hp\Documents\Data Science\Naresh IT\EDA with Python\Visadataset - Visadataset.csv")
visa_df.drop('case_id',axis=1,inplace=True)
pd.get_dummies(visa_df,dtype='int')

no_of_employees  yr_of_estab  prevailing_wage  continent_Africa  \
0                14513         2007         592.2029                 0   
1                 2412         2002       83425.6500                 0   
2                44444         2008      122996.8600                 0   
3                   98         1897       83434.0300                 0   
4                 1082         2005      149907.3900                 1   
...                ...          ...              ...               ...   
25475             2601         2008       77092.5700                 0   
25476             3274         2006      279174.7900                 0   
25477             1121         1910      146298.8500                 0   
25478             1918         1887       86154.7700                 0   
25479             3195         1960       70876.9100                 0   

       continent_Asia  continent_Europe  continent_North America  \
0                   1                 0                        0   
1                   1                 0                        0   
2                   1                 0                        0   
3                   1                 0                        0   
4                   0                 0                        0   
...               ...               ...                      ...   
25475               1                 0                        0   
25476               1                 0                        0   
25477               1                 0                        0   
25478               1                 0                        0   
25479               1                 0                        0   

       continent_Oceania  continent_South America  \
0                      0                        0   
1                      0                        0   
2                      0                        0   
3                      0                        0   
4                      0                        0   
...                  ...                      ...   
25475                  0                        0   
25476                  0                        0   
25477                  0                        0   
25478                  0                        0   
25479                  0                        0   

       education_of_employee_Bachelor's  ...  region_of_employment_South  \
0                                     0  ...                           0   
1                                     0  ...                           0   
2                                     1  ...                           0   
3                                     1  ...                           0   
4                                     0  ...                           1   
...                                 ...  ...                         ...   
25475                                 1  ...                           1   
25476                                 0  ...                           0   
25477                                 0  ...                           1   
25478                                 0  ...                           0   
25479                                 1  ...                           0   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                              1                  1                   0   
1                              0                  0                   0   
2                              1                  0                   0   
3                              1                  0                   0   
4                              0                  0                   0   
...                          ...                ...                 ...   
25475                          0                  0                   0   
25476                          0                  0                   0   
25477                          0                  0                   0   
2

Note:

- If dtype=True, not given then Values will be False and True

- Generally we prefer Label Encoder
- One hot encoder when categorical data has less Unique Labels